botorch/models/transforms/outcome.py

```diff
@@ -823,3 +828,523 @@ def untransform_posterior(self, posterior: Posterior) -> TransformedPosterior:
+                    denominator = dedup_median_index + 0.5 * (
+                        unique_y[dedup_median_index] == y_median
+                    )
+                    rank_quantile = 0.5 * (ranks - 0.5) / denominator
+
+                    y_above_median_std = self._get_std_above_median(unique_y, y_median)
+
+                    # Apply transformation
+                    rank_ppf = (
+                        torch.erfinv(2 * rank_quantile - 1)
+                        * y_above_median_std
+                        * torch.sqrt(torch.tensor(2.0))
+                    )
+                    Y_transformed[*batch_idx, :, dim] = torch.where(
+                        y < y_median,
+                        rank_ppf + y_median,
+                        Y_transformed[*batch_idx, :, dim],
+                    )
+
+                    # save intermediate values for untransform
+                    self._original_label_medians[*batch_idx, dim] = y_median
+                    self._unique_labels[(*batch_idx, dim)] = unique_y
+                    self._warped_labels[(*batch_idx, dim)] = (rank_ppf + y_median)[
+                        is_finite_mask
+                    ][unique_indices]
+
+            self._is_trained = torch.tensor(True)
+
+        return Y_transformed, Yvar
+
+    def untransform(
+        self, Y: Tensor, Yvar: Tensor | None = None
+    ) -> tuple[Tensor, Tensor | None]:
+        """Un-transform the outcomes.
+
+        Args:
+            Y: A `batch_shape x n x m`-dim tensor of transformed targets.
+            Yvar: A `batch_shape x n x m`-dim tensor of transformed observation
+                noises associated with the targets (if applicable).
+
+        Returns:
+            A two-tuple with the un-transformed outcomes:
+            - The un-transformed outcome observations.
+            - The un-transformed observation noise (if applicable).
+        """
+        if not self._is_trained:
+            raise RuntimeError("forward() needs to be called before untransform()")
+
+        if Yvar is not None:
+            raise NotImplementedError(
+                "HalfRankTransform does not support untransforming observation noise"
+            )
+
+        Y_utf = Y.clone()
+
+        for dim in range(Y.shape[-1]):
+            batch_indices = (
+                product(*(range(n) for n in self._batch_shape))
+                if len(self._batch_shape) > 0
+                else [  # this allows it to work with no batch dim
+                    (...,),
+                ]
+            )
+            for batch_idx in batch_indices:
+                y = Y_utf[*batch_idx, :, dim].clone()
+                unique_labels = self._unique_labels[(*batch_idx, dim)]
+                warped_labels = self._warped_labels[(*batch_idx, dim)]
+
+                # Process values below median
+                below_median = y < self._original_label_medians[*batch_idx, dim]
+                if below_median.any():
```

Comment:
This is quite a complicated process to untransform. I find it quite difficult to review since I don't know what the algorithm is supposed to do.

Reply:
The HalfRankTransform splits the data at the median and then warps the bottom half onto a gaussian whose std matches the std of the top half of the distribution. On inverting there are 3 cases, a) we can just look it up as it's very close to a known value from the fit, b) we take a in support linear interpolation, c) outside of the support we take a linear extrapolation based on the entire data range.